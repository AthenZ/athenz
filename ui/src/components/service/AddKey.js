/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import AddKeyForm from './AddKeyForm';
import ServiceKeyUtils from '../utils/ServiceKeyUtils';
import styled from '@emotion/styled';
import { colors } from '../denali/styles';
import Button from '../denali/Button';
import Color from '../denali/Color';
import RequestUtils from '../utils/RequestUtils';

const SectionsDiv = styled.div`
    width: 100%;
    text-align: left;
    background-color: ${colors.white};
`;

const ModifiedButton = styled(Button)`
    min-width: 8.5em;
    min-height: 1em;
`;

const ButtonDiv = styled.div`
    margin-left: 155px;
`;

const ParentButtonsDiv = styled.div`
    background-color: ${colors.white};
`;

const ErrorDiv = styled.div`
    margin-left: 155px;
`;

export default class AddKey extends React.Component {
    constructor(props) {
        super(props);
        this.api = this.props.api;
        this.onSubmit = this.onSubmit.bind(this);
        this.onChange = this.onChange.bind(this);
        this.state = {};
    }

    onSubmit() {
        if (!this.state.keyId || this.state.keyId === '') {
            this.setState({
                errorMessage: 'Key Id is required.',
            });
            return;
        }

        if (!this.state.keyValue || this.state.keyValue === '') {
            this.setState({
                errorMessage: 'Key Value is required.',
            });
            return;
        }

        this.api
            .addKey(
                this.props.domain,
                this.props.service,
                this.state.keyId,
                ServiceKeyUtils.y64Encode(
                    ServiceKeyUtils.trimKey(this.state.keyValue)
                ),
                this.props._csrf
            )
            .then(() => {
                this.props.onSubmit(
                    `Successfully created key with key id ${this.state.keyId} for service ${this.props.service}`
                );
            })
            .catch((err) => {
                this.setState({
                    errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                });
            });
    }

    onChange(key, value) {
        this.setState({ [key]: value });
    }

    render() {
        return (
            <SectionsDiv autoComplete={'off'} data-testid='add-key'>
                <AddKeyForm
                    api={this.api}
                    domain={this.props.domain}
                    onChange={this.onChange}
                />
                {this.state.errorMessage && (
                    <ErrorDiv>
                        <Color name={'red600'}>{this.state.errorMessage}</Color>
                    </ErrorDiv>
                )}
                <ParentButtonsDiv>
                    <ButtonDiv>
                        <ModifiedButton onClick={this.onSubmit}>
                            Submit
                        </ModifiedButton>
                        <ModifiedButton secondary onClick={this.props.cancel}>
                            Cancel
                        </ModifiedButton>
                    </ButtonDiv>
                </ParentButtonsDiv>
            </SectionsDiv>
        );
    }
}
