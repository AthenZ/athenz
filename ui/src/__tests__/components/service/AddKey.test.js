/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import AddKey from '../../../components/service/AddKey';

describe('AddKey', () => {
    it('should render', () => {
        const cancel = function() {};
        const domain = 'domain';
        const api = {};
        const { getByTestId } = render(
            <AddKey cancel={cancel} domain={domain} api={api} />
        );
        const addKey = getByTestId('add-key');
        expect(addKey).toMatchSnapshot();
    });

    it('should render nothing after cancel', async () => {
        const api = {
            addKey: function(domainName, serviceName, keyId, keyValue, _csrf) {
                return new Promise((resolve, reject) => {
                    resolve([]);
                });
            },
        };
        const cancel = function() {};
        const domain = 'domain';
        const { getByText, getByTestId, getByTitle } = render(
            <AddKey cancel={cancel} domain={domain} api={api} />
        );
        fireEvent.click(getByText('Cancel'));
        expect(
            await waitForElement(() => getByTestId('add-key'))
        ).toMatchSnapshot();
    });

    it('should render error after submit without keyID', async () => {
        const api = {
            addKey: function(domainName, serviceName, keyId, keyValue, _csrf) {
                return new Promise((resolve, reject) => {
                    resolve([]);
                });
            },
        };
        const cancel = function() {};
        const domain = 'domain';
        const { getByText, getByTestId, getByTitle } = render(
            <AddKey cancel={cancel} domain={domain} api={api} />
        );
        fireEvent.click(getByText('Submit'));
        expect(
            await waitForElement(() => getByTestId('error-message'))
        ).toMatchSnapshot();
    });

    it('should render error after submit without keyValue', async () => {
        const api = {
            addKey: function(domainName, serviceName, keyId, keyValue, _csrf) {
                return new Promise((resolve, reject) => {
                    resolve([]);
                });
            },
        };
        const cancel = function() {};
        const domain = 'domain';
        const { getByText, getByTestId, getByTitle } = render(
            <AddKey cancel={cancel} domain={domain} api={api} />
        );
        fireEvent.change(getByTestId('input-node'), {
            target: {
                value: 'test-id',
            },
        });

        await waitForElement(() => fireEvent.click(getByText('Submit')));
        expect(
            await waitForElement(() => getByTestId('error-message'))
        ).toMatchSnapshot();
    });

    it('should render error after submit throws error(refresh)', async () => {
        const api = {
            addKey: function(domainName, serviceName, keyId, keyValue, _csrf) {
                return new Promise((resolve, reject) => {
                    reject({
                        statusCode: 0,
                    });
                });
            },
        };
        const cancel = function() {};
        const domain = 'domain';
        const { getByText, getByTestId, getByTitle } = render(
            <AddKey cancel={cancel} domain={domain} api={api} />
        );
        fireEvent.change(getByTestId('input-node'), {
            target: {
                value: 'test-id',
            },
        });

        fireEvent.change(getByTestId('textarea'), {
            target: {
                value: 'test-value',
            },
        });

        await waitForElement(() => fireEvent.click(getByText('Submit')));
        expect(
            await waitForElement(() => getByTestId('error-message'))
        ).toMatchSnapshot();
    });

    it('should render error after submit throws error(other)', async () => {
        const api = {
            addKey: function(domainName, serviceName, keyId, keyValue, _csrf) {
                return new Promise((resolve, reject) => {
                    reject({
                        statusCode: 1,
                        body: {
                            message: 'test-error',
                        },
                    });
                });
            },
        };
        const cancel = function() {};
        const domain = 'domain';
        const { getByText, getByTestId, getByTitle } = render(
            <AddKey cancel={cancel} domain={domain} api={api} />
        );
        fireEvent.change(getByTestId('input-node'), {
            target: {
                value: 'test-id',
            },
        });

        fireEvent.change(getByTestId('textarea'), {
            target: {
                value: 'test-value',
            },
        });

        await waitForElement(() => fireEvent.click(getByText('Submit')));
        expect(
            await waitForElement(() => getByTestId('error-message'))
        ).toMatchSnapshot();
    });

    it('should trigger onSubmit after submit succeeds', async () => {
        const api = {
            addKey: function(domainName, serviceName, keyId, keyValue, _csrf) {
                return new Promise((resolve, reject) => {
                    resolve([]);
                });
            },
        };
        let test = 1;
        const submit = function() {
            test = 2;
        };
        const cancel = function() {};
        const domain = 'domain';
        const { getByText, getByTestId, getByTitle } = render(
            <AddKey
                cancel={cancel}
                domain={domain}
                api={api}
                onSubmit={submit}
            />
        );
        fireEvent.change(getByTestId('input-node'), {
            target: {
                value: 'test-id',
            },
        });

        fireEvent.change(getByTestId('textarea'), {
            target: {
                value: 'test-value',
            },
        });

        await waitForElement(() => fireEvent.click(getByText('Submit')));
        expect(await waitForElement(() => test)).toEqual(2);
    });
});
