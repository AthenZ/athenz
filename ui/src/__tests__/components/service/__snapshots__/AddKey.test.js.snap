// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddKey should render 1`] = `
<div
  autocomplete="off"
  class="css-vcd2ur"
  data-testid="add-key"
>
  <div
    class="css-y97ykc"
    data-testid="add-key-id"
  >
    <label
      class="css-fc0mv0-input-label denali-input-label css-9xcde0"
      for="key-id"
    >
      Public Key Id
    </label>
    <div
      class="css-1xqce5"
    >
      <div
        class="css-dqae60-input denali-input animated css-fvi5z5"
        data-testid="input-wrapper"
      >
        <input
          autocomplete="off"
          data-testid="input-node"
          id="key-id"
          name="key-id"
          value=""
        />
        <div
          class="message"
          data-testid="message"
        />
      </div>
    </div>
  </div>
  <div
    class="css-y97ykc"
    data-testid="add-key-value"
  >
    <label
      class="css-fc0mv0-input-label denali-input-label css-9xcde0"
      for="key-value"
    >
      Key Value
    </label>
    <div
      class="css-1xqce5"
    >
      <div
        class="css-1bg1ido-textarea denali-textarea css-6s8cn2"
        data-testid="textarea-wrapper"
      >
        <textarea
          data-testid="textarea"
          id="key-value"
          name="key-value"
        />
      </div>
    </div>
  </div>
  <div
    class="css-15tqd4u"
  >
    <div
      class="css-8y931m"
    >
      <button
        class="css-1iukreu-button denali-button css-7y8rgk"
      >
        Submit
      </button>
      <button
        class="css-1pobclq-button denali-button css-7y8rgk"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;

exports[`AddKey should render error after submit throws error(other) 1`] = `
<div
  data-testid="error-message"
  name="red600"
  style="color: rgb(208, 17, 17); display: inline-block;"
>
  Status: 1. Message: test-error
</div>
`;

exports[`AddKey should render error after submit throws error(refresh) 1`] = `
<div
  data-testid="error-message"
  name="red600"
  style="color: rgb(208, 17, 17); display: inline-block;"
>
  Session expired. Please refresh the page.
</div>
`;

exports[`AddKey should render error after submit without keyID 1`] = `
<div
  data-testid="error-message"
  name="red600"
  style="color: rgb(208, 17, 17); display: inline-block;"
>
  Key Id is required.
</div>
`;

exports[`AddKey should render error after submit without keyValue 1`] = `
<div
  data-testid="error-message"
  name="red600"
  style="color: rgb(208, 17, 17); display: inline-block;"
>
  Key Value is required.
</div>
`;

exports[`AddKey should render nothing after cancel 1`] = `
<div
  autocomplete="off"
  class="css-vcd2ur"
  data-testid="add-key"
>
  <div
    class="css-y97ykc"
    data-testid="add-key-id"
  >
    <label
      class="css-fc0mv0-input-label denali-input-label css-9xcde0"
      for="key-id"
    >
      Public Key Id
    </label>
    <div
      class="css-1xqce5"
    >
      <div
        class="css-dqae60-input denali-input animated css-fvi5z5"
        data-testid="input-wrapper"
      >
        <input
          autocomplete="off"
          data-testid="input-node"
          id="key-id"
          name="key-id"
          value=""
        />
        <div
          class="message"
          data-testid="message"
        />
      </div>
    </div>
  </div>
  <div
    class="css-y97ykc"
    data-testid="add-key-value"
  >
    <label
      class="css-fc0mv0-input-label denali-input-label css-9xcde0"
      for="key-value"
    >
      Key Value
    </label>
    <div
      class="css-1xqce5"
    >
      <div
        class="css-1bg1ido-textarea denali-textarea css-6s8cn2"
        data-testid="textarea-wrapper"
      >
        <textarea
          data-testid="textarea"
          id="key-value"
          name="key-value"
        />
      </div>
    </div>
  </div>
  <div
    class="css-15tqd4u"
  >
    <div
      class="css-8y931m"
    >
      <button
        class="css-1iukreu-button denali-button css-7y8rgk"
      >
        Submit
      </button>
      <button
        class="css-1pobclq-button denali-button css-7y8rgk"
      >
        Cancel
      </button>
    </div>
  </div>
</div>
`;
